Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx

```diff
@@ -314,10 +329,180 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+  torsionArray->SetNumberOfTuples(numberOfPoints);
+  torsionArray->Reset();
+  torsionArray->FillComponent(0,0.0);
+
+  // Initialize curvature variables
+  double minTorsion= 0.0;
+  double maxTorsion = 0.0;
+  double meanTorsion = 0.0;
+
+  // Get values for first point
+  double* binormal = binormals->GetTuple3(linePoints->GetId(1));
+  double binormalNorm = sqrt(binormal[0]*binormal[0] + binormal[1]*binormal[1] + binormal[2]*binormal[2]);
+  double normBinormal[3] = {0.0, 0.0, 0.0};
+  double prevNormBinormal[3] = {binormal[0]/binormalNorm, binormal[1]/binormalNorm, binormal[2]/binormalNorm};
+  double torsion = 0.0;
+  double *tangent = nullptr;
+  double currentLength = 0.0;
+  double length = 0.0;
+
+  // The torsion for the first cell is 0.0 for open curves
+  torsionArray->InsertValue(linePoints->GetId(0), 0.0);
+
+  for (vtkIdType idx=1; idx<numberOfPoints-1; ++idx)
+    {
+    binormal = binormals->GetTuple3(linePoints->GetId(idx+1));
+    binormalNorm = sqrt(binormal[0]*binormal[0] + binormal[1]*binormal[1] + binormal[2]*binormal[2]);
+
+    normBinormal[0] = binormal[0] / binormalNorm;
+    normBinormal[1] = binormal[1] / binormalNorm;
+    normBinormal[2] = binormal[2] / binormalNorm;
+
+    // Local torsion
+    torsion = sqrt( (normBinormal[0]-prevNormBinormal[0])*(normBinormal[0]-prevNormBinormal[0])
+                  + (normBinormal[1]-prevNormBinormal[1])*(normBinormal[1]-prevNormBinormal[1])
+                  + (normBinormal[2]-prevNormBinormal[2])*(normBinormal[2]-prevNormBinormal[2]) )
+              / binormalNorm;
+    torsionArray->InsertValue(linePoints->GetId(idx), torsion);
+
+    // Statistics
+    tangent = tangents->GetTuple3(linePoints->GetId(idx));
+    currentLength = sqrt( tangent[0]*tangent[0] + tangent[1]*tangent[1] + tangent[2]*tangent[2] );
+    if (torsion < minTorsion)
+      {
+      minTorsion = torsion;
+      }
+    else if (torsion > maxTorsion)
+      {
+      maxTorsion = torsion;
+      }
+    meanTorsion += torsion * currentLength; // weighted mean
+    length += currentLength;
+
+    // Propagate current values to previous
+    for (int i=0; i<3; ++i)
+      {
+      prevNormBinormal[i] = normBinormal[i];
+      }
+    } // For each line point
+
+  // Use the adjacent values for instead of the singular values
+  if (!this->CurveIsClosed)
+    {
+    torsionArray->InsertValue(linePoints->GetId(numberOfPoints-1),
+      torsionArray->GetValue(linePoints->GetId(numberOfPoints-2)));
+    }
+  else
+    {
+    torsionArray->SetComponent(linePoints->GetId(0), 0,
+      torsionArray->GetValue(linePoints->GetId(1)));
+    torsionArray->InsertValue(linePoints->GetId(numberOfPoints-1),
+      torsionArray->GetValue(linePoints->GetId(numberOfPoints-2)));
+    }
+
+  tangent = tangents->GetTuple3(linePoints->GetId(numberOfPoints-1));
+  currentLength = sqrt( tangent[0]*tangent[0] + tangent[1]*tangent[1] + tangent[2]*tangent[2] );
+  length += currentLength;
+  meanTorsion = meanTorsion / length;
```

Comment:
Length may be 0 (if points coincide), make sure that it does not cause major trouble.